File: sdks/typescript/trace-harness/src/runtime.ts
import { ConcurrencyPool } from "@compforge/harness-toolbox/concurrency";
import { randomUUID } from "node:crypto";
import { mkdir, mkdtemp, open, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import type { TraceHarness } from "./harness";
import type { Source, SpanQuery } from "./ingest/sources/base";
import { Dataset } from "./dataset";
import { AnalysisContext } from "./analyze/context";
import { diagnoseAnalysis } from "./analyze/diagnose";
import { assemble } from "./ingest/assemble";
import { TraceAnalysis } from "./loading/analysis";
import { EvidenceLoader } from "./loading/loader";
import { EvidenceStore } from "./loading/store";
import { EvidenceTooLarge, loadConfig, type LoadConfig } from "./loading/model";
import { STRUCTURE_FIELDS } from "./loading/structure";

export interface SessionOptions { workDir?: string; config?: LoadConfig }
function deferred<T>() {
  let resolve!: (value: T) => void, reject!: (error: unknown) => void;
  const promise = new Promise<T>((yes, no) => { resolve = yes; reject = no; });
  return { promise, resolve, reject };
}

/** Use `await using lease = await session.tree(...)` or close in finally.
 * The analysis is valid only while its lease is active.
 */
export class TraceLease {
  #closing?: Promise<void>;
  constructor(readonly analysis: AnalysisContext, private readonly release: () => void) {}
  close(): Promise<void> {
    return this.#closing ??= this.analysis.runtime!.close().finally(this.release);
  }
  [Symbol.asyncDispose](): Promise<void> { return this.close(); }
}

/** @spec A session owns Source lifetime and aggregate reading/active-trace budgets.
 * @why Dataset caches survive individual leases; analysis tasks never survive a lease.
 */
export class TraceSession {
  readonly config: Required<LoadConfig>;
  readonly #abort = new AbortController();
  readonly #reads: ConcurrencyPool;
  readonly #slots: ConcurrencyPool;
  readonly #loaders = new Map<string, EvidenceLoader>();
  readonly #leases = new Set<TraceLease>();
  readonly #operations = new Set<Promise<unknown>>();
  #workDir?: Promise<string>;
  #closing?: Promise<void>;
  constructor(readonly harness: TraceHarness, readonly source: Source, readonly options: SessionOptions = {}) {
    this.config = loadConfig(options.config);
    this.#reads = new ConcurrencyPool(this.config.concurrency);
    this.#slots = new ConcurrencyPool(this.config.activeTraces);
    const ids = harness.measurers.map(item => item.spec.id);
    if (new Set(ids).size !== ids.length) throw new Error("duplicate measurement id");
  }
  #check(): void { this.#abort.signal.throwIfAborted(); }
  #track<T>(work: Promise<T>): Promise<T> {
    this.#operations.add(work);
    void work.finally(() => this.#operations.delete(work)).catch(() => {});
    return work;
  }
  workspace(): Promise<string> {
    this.#check();
    return this.#workDir ??= this.options.workDir
      ? mkdir(this.options.workDir, { recursive: true }).then(() => this.options.workDir!)
      : mkdtemp(join(tmpdir(), "trace-harness-"));
  }
  select(query: SpanQuery = {}): Promise<Dataset> {
    this.#check();
    return this.#track(this.#select(query));
  }
  async #select(query: SpanQuery): Promise<Dataset> {
    const limit = query.limit ?? 1000;
    if (!Number.isSafeInteger(limit) || limit < 1) throw new Error("selection limit must be a positive integer");
    const id = randomUUID(), path = join(await this.workspace(), "datasets", id);
    await mkdir(path, { recursive: true });
    const members = new EvidenceStore(join(path, "members"));
    const stream = await open(join(path, "members.jsonl"), "wx");
    let count = 0;
    try {
      // Selection uses the same Source I/O budget, and never acquires trees.
      await this.#reads.run(async () => {
        for await (const traceId of this.source.select({ ...query, limit }, this.#abort.signal)) {
          this.#check();
          if (!traceId || typeof traceId !== "string") throw new Error("Source returned an empty trace identity");
          if (await members.get(traceId)) continue;
          await members.put(traceId, traceId);
          await stream.write(`${JSON.stringify(traceId)}\n`);
          if (++count >= limit) break;
        }
      }, this.#abort.signal);
      this.#check();
      await stream.close();
      await writeFile(join(path, "manifest.json"), JSON.stringify({ id, source: this.source.namespace, count, query, limit_reached: count === limit }));
      return new Dataset(id, this.source.namespace, path, count);
    } catch (error) { await stream.close(); await rm(path, { recursive: true, force: true }); throw error; }
  }
  #loader(dataset: Dataset): EvidenceLoader {
    this.#check();
    if (dataset.source !== this.source.namespace) throw new Error("dataset belongs to a different source");
    const key = dataset.path;
    if (!this.#loaders.has(key)) this.#loaders.set(key, new EvidenceLoader(this.source,
      new EvidenceStore(join(dataset.path, "evidence-cache")), this.config, this.#reads, this.#abort.signal,
      this.harness.contributions.fieldAliases, this.harness.contributions.normalizeSpan));
    return this.#loaders.get(key)!;
  }
  tree(dataset: Dataset, traceId: string): Promise<TraceLease> {
    this.#check();
    const ready = deferred<TraceLease>(), released = deferred<void>();
    const operation = this.#slots.run(async () => {
      if (!await dataset.contains(traceId)) throw new Error(`trace is not a member of dataset: ${traceId}`);
      const loader = this.#loader(dataset);
      const fields = [...new Set([...STRUCTURE_FIELDS, ...(this.harness.contributions.structureFields ?? []),
        ...[...this.harness.specs].flatMap(spec => spec.structure_fields ?? [])])];
      const skeleton = await loader.skeleton(traceId, fields);
      const physicalSpanIds = [...skeleton.keys()];
      const spans = this.harness.contributions.prepareSpans?.(skeleton) ?? skeleton;
      const trace = assemble(spans, this.harness.specs, this.harness.transforms, false, traceId);
      const runtime = new TraceAnalysis(this.harness, trace, loader, physicalSpanIds);
      const analysis = new AnalysisContext(trace, runtime.measurements, {}, runtime);
      const lease = new TraceLease(analysis, () => { this.#leases.delete(lease); released.resolve(); });
      this.#leases.add(lease);
      try {
        if (!this.config.lazy) {
          try { await loader.load(trace, physicalSpanIds, this.config.fields); }
          catch (error) {
            this.#check();
            // A failed unused field does not invalidate independent analysis; budgets always do.
            if (error instanceof EvidenceTooLarge) throw error;
          }
        }
        this.#check(); ready.resolve(lease);
        await released.promise;
      } catch (error) { await lease.close(); throw error; }
    }, this.#abort.signal);
    void this.#track(operation).catch(ready.reject);
    return ready.promise;
  }
  async *trees(dataset: Dataset): AsyncIterable<AnalysisContext> {
    for await (const id of dataset.members()) {
      const lease = await this.tree(dataset, id);
      try { yield lease.analysis; } finally { await lease.close(); }
    }
  }
  async analyze(analysis: AnalysisContext, options: { metrics?: readonly string[]; diagnosis?: boolean } = {}): Promise<AnalysisContext> {
    this.#runtime(analysis);
    const names = options.metrics ?? this.harness.measurers.map(item => item.spec.id);
    for (const name of names) if (!this.harness.measurers.some(item => item.spec.id === name)) throw new Error(`unknown measurement: ${name}`);
    const anchor = analysis.trace.nodes[0];
    if (anchor) await Promise.all(names.map(name => analysis.measure(anchor, name)));
    if (options.diagnosis === false) return analysis;
    await Promise.all(analysis.trace.nodes.flatMap(node => Object.keys(analysis.trace.specs.get(node.kind)?.metrics ?? {})
      .map(name => analysis.fact(node, name))));
    return diagnoseAnalysis(analysis, this.harness.detectors);
  }
  async prepareView(analysis: AnalysisContext, options: { full?: boolean } = {}): Promise<AnalysisContext> {
    await this.#runtime(analysis).prepareView(options.full); return analysis;
  }
  #runtime(analysis: AnalysisContext): TraceAnalysis {
    this.#check();
    if (!analysis.runtime || ![...this.#leases].some(lease => lease.analysis.runtime === analysis.runtime)) throw new Error("analysis does not belong to an active lease of this session");
    analysis.runtime.check(); return analysis.runtime;
  }
  get loadingStats() {
    return [...this.#loaders.values()].reduce((total, loader) => ({
      reads: total.reads + loader.stats.reads, fetches: total.fetches + loader.stats.fetches,
      bytesRead: total.bytesRead + loader.stats.bytesRead, cacheHits: total.cacheHits + loader.stats.cacheHits,
    }), { reads: 0, fetches: 0, bytesRead: 0, cacheHits: 0 });
  }
  close(): Promise<void> { return this.#closing ??= this.#close(); }
  async #close(): Promise<void> {
    this.#abort.abort(new Error("trace session is closed"));
    try {
      await Promise.allSettled([...this.#leases].map(lease => lease.close()));
      await Promise.allSettled([...this.#operations]);
      await Promise.allSettled([...this.#loaders.values()].map(loader => loader.close()));
      await this.source.close();
    } finally {
      this.#loaders.clear();
      if (!this.options.workDir && this.#workDir) await rm(await this.#workDir, { recursive: true, force: true });
    }
  }
  [Symbol.asyncDispose](): Promise<void> { return this.close(); }
}

File: sdks/python/trace_harness/runtime.py
"""One managed execution path for fixed datasets, including single-member datasets."""

from __future__ import annotations

import asyncio
import json
import shutil
import sqlite3
import tempfile
from collections.abc import AsyncIterator, Awaitable
from contextlib import asynccontextmanager
from dataclasses import asdict
from pathlib import Path
from typing import TYPE_CHECKING
from uuid import uuid4

from trace_harness.ingest.sources.base import Source

if TYPE_CHECKING:
    from trace_harness.batch import BatchResult
    from trace_harness.harness import TraceHarness

from trace_harness.analyze.context import AnalysisContext
from trace_harness.analyze.diagnose import diagnose
from trace_harness.analyze.diagnose.registry import DetectorRegistry
from trace_harness.dataset import Dataset
from trace_harness.ingest.assemble import assemble
from trace_harness.ingest.sources.base import SpanQuery
from trace_harness.loading.analysis import TraceAnalysis
from trace_harness.loading.loader import EvidenceLoader
from trace_harness.loading.model import EvidenceTooLarge, LoadConfig
from trace_harness.loading.store import EvidenceStore

# Generic structural and inexpensive quantitative metadata. Business additions are explicit.
STRUCTURE_FIELDS = (
    "asgi.event.type",
    "http.target",
    "server.address",
    "server.port",
    "net.peer.name",
    "net.peer.port",
    "span.kind",
    "http.url",
    "url.full",
    "url.path",
    "http.method",
    "http.request.method",
    "http.status_code",
    "http.response.status_code",
    "http.route",
    "otel.status_code",
    "error",
    "otel.status_description",
    "gen_ai.operation.name",
    "gen_ai.request.model",
    "gen_ai.response.model",
    "llm.model_name",
    "llm.response.model",
    "gen_ai.tool.name",
    "gen_ai.agent.name",
    "gen_ai.agent.id",
    "gen_ai.usage.input_tokens",
    "gen_ai.usage.output_tokens",
    "gen_ai.usage.total_tokens",
    "gen_ai.usage.prompt_tokens",
    "gen_ai.usage.completion_tokens",
    "llm.token_count.prompt",
    "llm.token_count.completion",
    "llm.token_count.total",
    "http.request_id",
    "request.id",
)


class TraceSession:
    def __init__(
        self,
        harness: TraceHarness,
        source: Source,
        *,
        work_dir: str | Path | None = None,
        config: LoadConfig | None = None,
    ):
        self.harness, self.source = harness, source
        self.config = config or LoadConfig()
        self._temp = (
            tempfile.TemporaryDirectory(prefix="trace-harness-") if work_dir is None else None
        )
        self.work_dir = Path(self._temp.name if self._temp else work_dir)
        self._stores = {}
        self._loaders = {}
        self._slots = asyncio.Semaphore(self.config.active_traces)
        self._closed = False
        self._leases = set()

    async def __aenter__(self):
        return self

    async def __aexit__(self, *exc):
        self._closed = True
        for lease in tuple(self._leases):
            await lease.close()
        for loader in self._loaders.values():
            await loader.close()
        for store in self._stores.values():
            store.close()
        await self.source.aclose()
        if self._temp:
            self._temp.cleanup()

    async def select(self, query: SpanQuery | None = None) -> Dataset:
        if self._closed:
            raise RuntimeError("session is closed")
        query = query or SpanQuery()
        dataset_id = uuid4().hex
        path = self.work_dir / "datasets" / dataset_id
        path.mkdir(parents=True)
        count = 0
        members_db = sqlite3.connect(path / "index.sqlite")
        try:
            members_db.execute("CREATE TABLE members(trace TEXT PRIMARY KEY)")
            with (path / "members.jsonl.tmp").open("w") as stream:
                async for trace_id in self.source.select(query):
                    if not members_db.execute(
                        "INSERT OR IGNORE INTO members VALUES(?)", (trace_id,)
                    ).rowcount:
                        continue
                    stream.write(json.dumps(trace_id) + "\n")
                    count += 1
            members_db.commit()
        except BaseException:
            members_db.close()
            shutil.rmtree(path)
            raise
        finally:
            members_db.close()
        (path / "members.jsonl.tmp").replace(path / "members.jsonl")
        (path / "manifest.json").write_text(
            json.dumps(
                {
                    "id": dataset_id,
                    "source": self.source.namespace,
                    "query": asdict(query),
                    "count": count,
                    "limit_reached": count >= query.limit,
                },
                indent=2,
            )
        )
        return Dataset(dataset_id, self.source.namespace, path)

    def _loader(self, dataset: Dataset) -> EvidenceLoader:
        if self._closed:
            raise RuntimeError("session is closed")
        if dataset.source != self.source.namespace:
            raise ValueError("dataset belongs to a different source")
        if dataset.id not in self._loaders:
            store = EvidenceStore(dataset.path)
            self._stores[dataset.id] = store
            self._loaders[dataset.id] = EvidenceLoader(
                self.source,
                store,
                self.config,
                field_aliases=self.harness.contributions.field_aliases,
                normalize_span=self.harness.contributions.normalize_span,
            )
        return self._loaders[dataset.id]

    @asynccontextmanager
    async def tree(self, dataset: Dataset, trace_id: str) -> AsyncIterator[AnalysisContext]:
        if not dataset.contains(trace_id):
            raise KeyError(f"trace is not a member of dataset: {trace_id}")
        async with self._slots:
            loader = self._loader(dataset)
            fields = tuple(
                sorted(
                    set(STRUCTURE_FIELDS).union(
                        self.harness.contributions.structure_fields,
                        *(spec.structure_fields for spec in self.harness.specs),
                    )
                )
            )
            spans = await loader.skeleton(trace_id, fields)
            if self.harness.contributions.prepare_spans:
                spans = self.harness.contributions.prepare_spans(spans)
            trace = assemble(
                spans, self.harness.specs, transforms=self.harness.transforms, prepare=False
            )
            trace.trace_id = trace_id
            trace.evidence_dir = dataset.path / "evidence" / trace_id
            runtime = TraceAnalysis(self.harness, trace, loader)
            self._leases.add(runtime)
            try:
                if not self.config.lazy:
                    # Preload errors are recorded by the loader and surface only if consumed.
                    try:
                        await loader.load(
                            trace,
                            tuple(sid for sid, s in trace.spans.items() if s.raw.get("spanID")),
                            self.config.fields,
                        )
                    except EvidenceTooLarge:
                        # Resource budgets apply even when preload evidence is not consumed.
                        raise
                    except Exception:
                        pass  # The loader retains per-field errors for subsequent consumers.
                yield AnalysisContext(trace, runtime.measurements, {}, runtime)
            finally:
                await runtime.close()
                self._leases.discard(runtime)

    async def trees(self, dataset: Dataset) -> AsyncIterator[AnalysisContext]:
        for trace_id in dataset.members():
            async with self.tree(dataset, trace_id) as analysis:
                yield analysis

    async def analyze(
        self,
        analysis: AnalysisContext,
        *,
        detectors: list[str] | None = None,
        metrics: list[str] | None = None,
        probes: bool = False,
    ) -> AnalysisContext:
        metric_ids = [m.spec.id for m in self.harness.measurers] if metrics is None else metrics
        if analysis.trace.nodes:
            await asyncio.gather(
                *(analysis.measure(analysis.trace.nodes[0], name) for name in metric_ids)
            )
        if detectors is not None:
            registered = self.harness.detectors.registered()
            chosen = [d for d in registered if d.__name__ in detectors]
            unknown = set(detectors) - {d.__name__ for d in chosen}
            if unknown:
                raise KeyError(f"unknown detectors: {sorted(unknown)}")
            # An explicit detector selection does not run unrelated baseline rules.
            found = {}
            from trace_harness.analyze.diagnose import _post_order

            for node in _post_order(analysis.trace):
                for detector in chosen:
                    result = detector(node, analysis)
                    for finding in (
                        await result if isinstance(result, Awaitable) else result
                    ) or []:
                        found.setdefault(finding.node_id, []).append(finding)
        else:
            # Kind metrics may themselves depend on evidence (e.g. tool result size).
            await asyncio.gather(
                *(
                    analysis.fact(n, name)
                    for n in analysis.trace.nodes
                    for name in analysis.trace.specs[n.kind].metrics
                )
            )
            found = await diagnose(
                analysis.trace,
                probes=probes,
                detector_registry=DetectorRegistry(self.harness.detectors.registered()),
                analysis=analysis,
            )
        return AnalysisContext(analysis.trace, analysis.measurements, found, analysis.runtime)

    async def prepare_view(
        self, analysis: AnalysisContext, *, full: bool = False
    ) -> AnalysisContext:
        if analysis.runtime is None:
            raise ValueError("view preparation requires an active trace lease")
        await analysis.runtime.prepare_view(full=full)
        return analysis

    async def detect(
        self,
        dataset: Dataset,
        *,
        detectors: list[str] | None = None,
        metrics: list[str] | None = None,
        batch_detectors: list[str] | None = None,
    ) -> BatchResult:
        from trace_harness.batch import run_batch

        return await run_batch(
            self, dataset, detectors=detectors, metrics=metrics, batch_detectors=batch_detectors
        )
